src/compiler/scala/tools/nsc/typechecker/Namers.scala

```diff
@@ -1070,7 +1070,16 @@ trait Namers extends MethodSynthesis {
 
       val defnTpe = widenIfNecessary(tree.symbol, rhsTpe, pt)
       tree.tpt defineType defnTpe setPos tree.pos.focus
-      tree.tpt.tpe
+      val tpe = tree.tpt.tpe
+      // if enabled, validate that the now inferred val or def type isn't PwS
+      if (settings.warnInferProductWithSerializable && context.reportErrors) {
+        tpe match {
+          case RefinedType(ProductRootTpe :: SerializableTpe :: _, _) =>
```

Comment:
Possibly check that `scope.isEmpty` here? I'm unsure how it would really make a difference, though.